File: java/src/main/java/problems101_200/PascalsTriangle2.java
package problems101_200;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

enum PascalsTriangle2 {
    DEFAULT {
        @Override
        List<Integer> getRow(int rowIndex) {
            int numRows = rowIndex + 1;

            List<List<Integer>> rows = new ArrayList<>(numRows);
            rows.add(Collections.singletonList(1));
            for (int i = 1; i < numRows; i++) {
                int numColumns = i + 1;
                List<Integer> columns = new ArrayList<>(numColumns);
                for (int j = 0; j < numColumns; j++) {
                    if (j == 0 || j == numColumns - 1) {
                        columns.add(1);
                        continue;
                    }
                    List<Integer> prevColumns = rows.get(i - 1);
                    columns.add(prevColumns.get(j - 1) + prevColumns.get(j));
                }
                rows.add(columns);
            }
            return rows.get(rows.size() - 1);
        }
    };

    abstract List<Integer> getRow(int rowIndex);
}

File: kotlin/src/main/kotlin/problems101_200/PascalsTriangle2.kt
package problems101_200

enum class PascalsTriangle2 {
    DEFAULT {
        override fun getRow(rowIndex: Int): List<Int> {
            val numRows = rowIndex + 1

            val rows = ArrayList<List<Int>>(numRows)
            rows.add(listOf(1))
            for (i in 1 until numRows) {
                val numColumns = i + 1
                val columns = ArrayList<Int>(numColumns)
                for (j in 0 until numColumns) {
                    if (j == 0 || j == numColumns - 1) {
                        columns.add(1)
                        continue
                    }
                    val prevColumns = rows[i - 1]
                    columns.add(prevColumns[j - 1] + prevColumns[j])
                }
                rows.add(columns)
            }
            return rows[rows.size - 1]
        }
    },
    ;

    abstract fun getRow(rowIndex: Int): List<Int>
}
